nimbleclient/v1/restclient.py

```diff
@@ -164,81 +168,93 @@ def close_connection(self):
+                        params=params,
+                        headers=self._headers,
+                        verify=False
+                    )
 
                 if response.status_code >= 400:
                     if 'SM_http_unauthorized' in str(response.content):
                         self._refresh_connection()
                     else:
-                        print("URL: ", url)
                         raise NimOSAPIError(response.json())
-                else:
+
+                response = response.json()
+                logging.debug("API RESPONSE: %s", json.dumps(response, indent=4))
+
+                # Check for errors if any in the response (Treat partial response as an error)
+                if 'messages' in response:
+                    raise NimOSAPIError(response['messages'])
+
+                # Get response
+                if 'totalRows' not in response:
+                    logging.debug("Returning Data: %s", json.dumps(response, indent=4))
+                    return response
+
+                # Append to the resultant data
+                response_data.extend(response["data"])
+
+                # When pageSize is requested, then return only those no. of records
+                if 'pageSize' in params:
+                    break
+
+                # When end_row is requested, then return only upto 'end_row' records
+                if 'endRow' in params and params['endRow'] == response['endRow']:
+                    break
+
+                # Break when all available records are read or No more records to read
+                if len(response_data) == response['totalRows'] or len(response["data"]) == 0:
                     break
 
-            # Check for errors if any in the response (Treat partial response as an error)
-            if 'messages' in response.json():
-                raise NimOSAPIError(response.json()['messages'])
-
-            # Retrieves as per 'rest_api_row_limit' configuration on array
-            if 'pageSize' in params:
-                return response.json()
-
-            if 'totalRows' not in response.json():
-                return response.json()
-
-            # If startRow and/or endRow is specified, then retrieve records accordingly.
-            # If unspecified, then retrieves all the available records.
-            paginated_data = list()
-            paginated_data.extend(response.json()["data"])
-            total_rows = response.json()['totalRows']
-            retrieved_rows = len(paginated_data)
-            #current_end_row = response.json()['endRow']
-
-            if 'startRow' in params and 'endRow' in params:
-                requested_rows = params['endRow'] - params['startRow']
-            elif 'startRow' not in params and 'endRow' not in params:
-                requested_rows = total_rows # Get all available records
-            else:
+                # Set next start row index to continue reading more records
                 if 'startRow' in params:
```

Comment:
If you're expecting startRow and other parameters that won't be part of the fields you pass into your request, then you should add them into the function signature and use snake_case so its consistent with the rest of the code: def get_data(self, endpoint, filter=None, start_row=None, end_row=None, page_size=None, **params):.

Reply:
But startRow, endRow, and pageSize are not snake-cased in NimOS.  That's a bug on their end that needs to be addressed.

Reply:
Ah ok.  Agreed.  We should handle it if we can.